mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1715,6 +1715,30 @@ pub async fn ordered_mature_unspents<T>(coin: T, address: Address) -> Result<Vec
 where
     T: AsRef<UtxoCoinFields> + UtxoCommonOps,
 {
+    fn calc_actual_tx_confirmations(tx: &RpcTransaction, block_count: u64) -> Result<u32, String> {
+        let tx_height = tx
```

Comment:
Does it make sense to return the `Result<u32, String>` from this fn if we just skip the error next? 
When `tx_height` is None then transaction is unconfirmed so we can return 0.

I don't remember, doesn't `tx_height == 0` mean that transaction is in genesis block? Or was there the case when RPC returned `0` for unconfirmed transaction?

`if block_count < tx_height` case is a bit tricky, but I guess that we can even return 1 here because it means that new block was mined of which we are not aware yet. Or maybe reset the block_count to tx_height.

Reply:
> Does it make sense to return the Result<u32, String> from this fn if we just skip the error next?

Yes, but we can just log an error and exit from the function

>  When tx_height is None then transaction is unconfirmed so we can return 0.

I think we should return the previous confirmations value instead of 0. I decided to avoid caching of transactions with `tx.height = 0`: it's extremely rare case when `list_unspent_ordered()` returns transactions in genesis block.
It means we don't cache transactions with `tx.height = 0` and we don't expect transactions with `tx.height = 0` from cache. Anyway if we get transactions in genesis block, we just request them using `get_verbose_transaction()`

> if block_count < tx_height case is a bit tricky, but I guess that we can even return 1 here because it means that new block was mined of which we are not aware yet

Is it possible to core/electrum can return unspent transactions and be unaware of the new block?

Reply:
I think it should be either:
1) `calc_actual_tx_confirmations` returns `Result<new_confirmations, String>` and we reset `tx.confirmations` on success
2) `calc_actual_tx_confirmations` returns `previous_confirmations` on error or `new_confirmations` on success

Reply:
> I think we should return the previous confirmations value instead of 0. I decided to avoid caching of transactions with tx.height = 0: it's extremely rare case when list_unspent_ordered() returns transactions in genesis block.

The relation between transaction caching and this fn confused me initially, but then I've understood that this fn is used only to calc confirmations for transactions saved in cache.

However the implementation detail of `cache` looks excess to me here.

Imagine if one day we decide to change the cache implementation and forget about this fn. We will get a bug as cache will always return outdated confirmations amount. Ofc there's very little probability of such event and it will be logged, but this `tx.height == 0` also interferes the `calc_actual_tx_confirmations` to become just a general way to calculate transaction confirmations.

Or it should be at least called `calc_actual_cached_tx_confirmations` to be more explicit.

Reply:
> Is it possible to core/electrum can return unspent transactions and be unaware of the new block?

Yes, this the actual reason of this commit :slightly_smiling_face: I'm unsure how Electrum indexing works in details, but I can easily imagine that at some point of time it can index UTXOs for some address and only then import the new block data returning outdated number in `blockchain.headers.subscribe`.

And it's also possible for e.g. `blockchain.transaction.get` - this call forwarded as is to underlying coin daemon so it's state can be a bit ahead of Electrum.

I also guess it's not likely or even impossible in native mode.

Reply:
I agree the `calc_actual_cached_tx_confirmations` is clearer than `calc_actual_tx_confirmations`. I will change it now, thanks :)

Reply:
Done

Reply:
Thanks! PR approved :slightly_smiling_face: